hub/api/pytorch.py

```diff
@@ -0,0 +1,241 @@
+import os
+import pickle
+import numpy as np
+from itertools import repeat
+from collections import defaultdict
+from hub.util.keys import get_index_map_key
+from hub.core.chunk_engine.chunker import join_chunks
+from hub.core.chunk_engine.read import read_tensor_meta
+from pathos.pools import ProcessPool
+from multiprocessing import resource_tracker
+from multiprocessing.shared_memory import SharedMemory
+from hub.util.exceptions import ModuleNotInstalledException
+
+
+# TODO make this use shared memory to make on the fly transforms faster. Currently using transform slows us down by 10x
+def apply_transform(transform, sample):
+    """Used to apply transforms to a single samples"""
+    return transform(sample) if transform else sample
+
+
+def shared_memory_clear(chunk_set):
+    """Helper function that checks if an existing SharedMemory exists for any chunk in chunk_set and clears it"""
+    for chunk_name in chunk_set:
+        try:
+            shm = SharedMemory(name=chunk_name)
+            shm.close()
+            shm.unlink()
+        except:
+            pass
+
+
+def read_chunk(chunk, key):
+    """Reads a single chunk from the dataset's storage provider and stores it in the SharedMemory"""
+    remove_shared_memory_from_resource_tracker()
+    chunk_path = os.path.join(key, "chunks", chunk)
+    chunk_bytes = _hub_storage_provider[chunk_path]
+    chunk = chunk_path.split("/")[-1]
+    shm = SharedMemory(create=True, size=len(chunk_bytes), name=chunk)
+    shm.buf[:] = chunk_bytes
+    shm.close()
+    return
+
+
+def remove_shared_memory_from_resource_tracker():
+    """Helper function to fix bug in Python SharedMemory
+
+    Monkey-patch multiprocessing.resource_tracker so SharedMemory won't be tracked
+    More details at: https://bugs.python.org/issue38119
+    """
+
+    def fix_register(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.register(self, name, rtype)
+
+    def fix_unregister(name, rtype):
+        if rtype == "shared_memory":
+            return
+        return resource_tracker._resource_tracker.unregister(self, name, rtype)
+
+    resource_tracker.register = fix_register
+    resource_tracker.unregister = fix_unregister
+    if "shared_memory" in resource_tracker._CLEANUP_FUNCS:
+        del resource_tracker._CLEANUP_FUNCS["shared_memory"]
+
+
+def _to_pytorch(dataset, transform=None, workers=1):
+    try:
+        import torch
+    except ModuleNotFoundError:
+        raise ModuleNotInstalledException("torch")
+    global torch
+    global _hub_storage_provider  # global to pass to processes, as not possible to serialize it and send as argument
+
+    # TODO, apparently boto3.client isn't safe for multiprocessing https://github.com/boto/boto3/pull/2848/files
+    # could it be working here as we're only reading data?
+    _hub_storage_provider = dataset.provider
+    return TorchDataset(dataset, transform, workers)
```

Comment:
does it make sense to move this whole function into `TorchDataset.__init__`? it seems like that's all it's doing